src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/TextEditResponseRewriter.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion.Delegation
+{
+    internal class TextEditResponseRewriter : DelegatedCompletionResponseRewriter
+    {
+        public override int Order => ExecutionBehaviorOrder.ChangesCompletionItems;
+
+        public override async Task<VSInternalCompletionList> RewriteAsync(
+            VSInternalCompletionList completionList,
+            int hostDocumentIndex,
+            DocumentContext hostDocumentContext,
+            DelegatedCompletionParams delegatedParameters,
+            CancellationToken cancellationToken)
+        {
+            if (delegatedParameters.ProjectedKind != RazorLanguageKind.CSharp)
+            {
+                return completionList;
+            }
+
+            var sourceText = await hostDocumentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+
+            sourceText.GetLineAndOffset(hostDocumentIndex, out var lineNumber, out var characterOffset);
+            var hostDocumentPosition = new Position(lineNumber, characterOffset);
+            completionList = TranslateTextEdits(hostDocumentPosition, delegatedParameters.ProjectedPosition, completionList);
+
+            if (completionList.ItemDefaults?.EditRange != null)
+            {
+                completionList.ItemDefaults.EditRange = TranslateRange(hostDocumentPosition, delegatedParameters.ProjectedPosition, completionList.ItemDefaults.EditRange);
+            }
+
+            return completionList;
+        }
+
+        // The TextEdit positions returned to us from the C#/HTML language servers are positions correlating to the virtual document.
```

Comment:
Nit: if the comment tells people what the method does, it should be a doc comment. If the comment is details about what the code inside the method is doing, it should be inside the method.

This feels like the latter to me.